ee/server/service/scripts.go

```diff
@@ -0,0 +1,216 @@
+			}
+			after.Reset(checkInterval)
+		}
+	}
+}
+
+func (svc *Service) GetScriptResult(ctx context.Context, execID string) (*fleet.HostScriptResult, error) {
+	scriptResult, err := svc.ds.GetHostScriptExecutionResult(ctx, execID)
+	if err != nil {
+		if fleet.IsNotFound(err) {
+			if err := svc.authz.Authorize(ctx, &fleet.HostScriptResult{}, fleet.ActionRead); err != nil {
+				return nil, err
+			}
+		}
+		svc.authz.SkipAuthorization(ctx)
+		return nil, ctxerr.Wrap(ctx, err, "get script result")
+	}
+
+	host, err := svc.ds.HostLite(ctx, scriptResult.HostID)
+	if err != nil {
+		// if error is because the host does not exist, check first if the user
+		// had access to run a script (to prevent leaking valid host ids).
+		if fleet.IsNotFound(err) {
+			if err := svc.authz.Authorize(ctx, &fleet.HostScriptResult{}, fleet.ActionRead); err != nil {
+				return nil, err
+			}
+		}
+		svc.authz.SkipAuthorization(ctx)
+		return nil, ctxerr.Wrap(ctx, err, "get host lite")
+	}
+	if err := svc.authz.Authorize(ctx, &fleet.HostScriptResult{TeamID: host.TeamID}, fleet.ActionRead); err != nil {
+		return nil, err
+	}
+
+	scriptResult.Hostname = host.DisplayName()
+
+	return scriptResult, nil
+}
+
+func (svc *Service) NewScript(ctx context.Context, teamID *uint, name string, r io.Reader) (*fleet.Script, error) {
+	if err := svc.authz.Authorize(ctx, &fleet.Script{TeamID: teamID}, fleet.ActionWrite); err != nil {
+		return nil, err
+	}
+
+	if name == "" {
+		return nil, fleet.NewInvalidArgumentError("script", "The file name must not be empty.")
+	}
+	if filepath.Ext(name) != ".sh" {
+		return nil, fleet.NewInvalidArgumentError("script", "The file should be a .sh file.")
+	}
+
+	b, err := io.ReadAll(r)
+	if err != nil {
+		return nil, ctxerr.Wrap(ctx, err, "read script contents")
+	}
+	contents := string(b)
+	if err := fleet.ValidateHostScriptContents(contents); err != nil {
+		return nil, fleet.NewInvalidArgumentError("script", err.Error())
+	}
+
+	script := &fleet.Script{
+		TeamID:         teamID,
+		Name:           name,
+		ScriptContents: contents,
+	}
+	savedScript, err := svc.ds.NewScript(ctx, script)
+	if err != nil {
+		var (
+			existsErr fleet.AlreadyExistsError
+			fkErr     fleet.ForeignKeyError
+		)
+		if errors.As(err, &existsErr) {
+			err = fleet.NewInvalidArgumentError("script", "A script with this name already exists.").WithStatus(http.StatusConflict)
+		} else if errors.As(err, &fkErr) {
+			err = fleet.NewInvalidArgumentError("team_id", "The team does not exist.").WithStatus(http.StatusNotFound)
+		}
+		return nil, ctxerr.Wrap(ctx, err, "create script")
+	}
+	return savedScript, nil
+}
+
+func (svc *Service) DeleteScript(ctx context.Context, scriptID uint) error {
+	script, err := svc.ds.Script(ctx, scriptID)
+	if err != nil {
+		if fleet.IsNotFound(err) {
+			// couldn't get the script to have its team, authorize with a no-team script
+			if err := svc.authz.Authorize(ctx, &fleet.Script{}, fleet.ActionWrite); err != nil {
```

Comment:
sanity checking my understanding: will I get a `404`? if I:

-  try to delete a team script
- using the wrong ID
- with the right permissions

Reply:
Right, so the thing is if you have the wrong script ID (as in a non-existent one), we have no way to know for what team that non-existing script would have been. So what it does instead of returning of 404 immediately, it checks if the user had authorization to delete a non-team script (it defaults non-existing scripts to a non-team one) and fails with an authorization check if that's the case to avoid leaking valid/invalid IDs otherwise (i.e. 404 if it doesn't exist, authorization failure if it does).

So to answer your question, for a non-existing script id, you'd get a 404 if (and only if) you have write-access to no-team scripts. 

The only way I can see this return 404 for team users would be if we required the team id to be explicitly provided along with the script id.

Reply:
that makes sense. Thank you!